src/error.rs

```diff
@@ -3,77 +3,61 @@
+    #[error("Prefix {0} is not allowed")]
     InvalidPrefix(&'static str),
-    #[fail(display = "Sequence {} is not allowed", _0)]
+    #[error("Sequence {0} is not allowed")]
     InvalidSequence(&'static str),
-    #[fail(display = "{}", _0)]
-    Http(#[fail(cause)] HttpError),
-    #[fail(display = "{}", _0)]
-    HttpStatus(#[fail(cause)] HttpStatusError),
-    #[fail(display = "An HTTP response didn't have a valid Content-Length")]
+    #[error("HTTP error")]
+    Http(#[source] HttpError),
+    #[error("HTTP status")]
+    HttpStatus(#[source] HttpStatusError),
+    #[error("An HTTP response didn't have a valid Content-Length")]
     UnknownContentLength,
-    #[fail(display = "GCS API error: {}", _0)]
-    API(#[fail(cause)] ApiError),
-    #[fail(display = "JSON error: {}", _0)]
-    Json(#[fail(cause)] JsonError),
-    #[fail(display = "Response body doesn't contain enough data")]
+    #[error("GCS API error")]
+    API(#[source] ApiError),
+    #[error("JSON error")]
+    Json(#[source] JsonError),
+    #[error("Response body doesn't contain enough data")]
     InsufficientData,
-    #[fail(display = "Key rejected: {}", _0)]
+    #[error("Key rejected: {0}")]
     KeyRejected(&'static str),
-    #[fail(display = "An error occurred during signing")]
+    #[error("An error occurred during signing")]
     SigningError,
-    #[fail(
-        display = "An expiration duration was too long: requested = {}, max = {}",
-        requested, max
-    )]
+    #[error("An expiration duration was too long: requested = {requested}, max = {max}")]
     TooLongExpiration { requested: u64, max: u64 },
-    #[fail(display = "Failed to parse url {}", _0)]
-    UrlParse(url::ParseError),
-    #[fail(display = "Unable to stringize header value '{:?}'", _0)]
+    #[error("Failed to parse url")]
+    UrlParse(#[source] url::ParseError),
+    #[error("Unable to stringize header value '{0:?}'")]
     OpaqueHeaderValue(http::header::HeaderValue),
-    #[fail(display = "I/O error occurred: {}", _0)]
-    Io(#[fail(cause)] IoError),
-    #[fail(display = "Unable to decode base64: {}", _0)]
-    Base64Decode(base64::DecodeError),
-    #[fail(display = "Unable to url encode: {}", _0)]
-    UrlEncode(serde_urlencoded::ser::Error),
+    #[error("I/O error occurred")]
+    Io(#[source] IoError),
+    #[error("Unable to decode base64")]
+    Base64Decode(#[source] base64::DecodeError),
+    #[error("Unable to encode url")]
+    UrlEncode(#[source] serde_urlencoded::ser::Error),
 }
 
-#[derive(Debug)]
-pub struct HttpError(http::Error);
-
-impl PartialEq for HttpError {
-    fn eq(&self, _other: &Self) -> bool {
-        // I feel really bad about this
-        true
-    }
-}
+#[derive(Debug, thiserror::Error)]
+pub struct HttpError(#[source] http::Error);
 
 impl fmt::Display for HttpError {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        std::error::Error::description(self).fmt(f)
+        write!(f, "{}", self.0)
```

Comment:
You should be able to get this whole impl by writing `#[error("{0}")]` on the struct.

Reply:
Ahh cool, this was just a first pass so I tried to change as little as possible. :slightly_smiling_face: